### ISSUE TITLE
Using engine thrusters space as a special reverse gun port

### ISSUE THREAD

**Issue by jostephd** [endless-sky/endless-sky#3776]
_edit See https://github.com/endless-sky/endless-sky/issues/3776#issuecomment-397336387 for a better explanation_

Reverse thrusters take up weapon space.

Couldn't this be done the other way, too?  That is, could one write a "Reverse Heavy Rocket Launcher" outfit that takes up engine space (rather than weapon space) and its missiles launch aft? The `Hardpoint` constructor doesn't take an Angle parameter so I guess that's not currently possible.

A hypothetical Reverse Heavy Rocket Launcher would be useful when trying to outrun fast opponents, since Heavy Rockets don't home.

**Comment by tehhowch** [comment]
Rear-firing is just one case of angled hardpoints (#2237).

You can make outfits take up engine space even if they are not engines already, so I think this is just a duplicate of #2237 and recommend closing it.

**Comment by jostephd** [comment]
Good point: If angled hardpoints exist, one could implement "reverse guns in engine space" by writing an outfit that consumes engine space and adds a rear-facing gunport. Would it be possible for that hypothetical outfit to specify the gunport's x,y? Or would that require a C++ change?

**Comment by jostephd** [comment]
Thanks for the very detailed answer! I'll change the title shortly. To clarify, I wasn't trying to say "I want this to be implemented", I was merely trying to brainstorm whether reverse guns would be a good addition.

**Comment by Amazinite** [comment]
>To clarify, I wasn't trying to say "I want this to be implemented", I was merely trying to brainstorm whether reverse guns would be a good addition.

The problems with reverse firing weapons are the same problems with broadside weapons and any other non-normally angled weapon. i.e. discussing reverse firing weapons will just be the same discussion as #2237. So as @tehhowch said, this is effectively a duplicate and therefore should probably be closed.

**Comment by jostephd** [comment]
@Amazinite As @tehhowch said in his later comment, there is a distinction between #2237 and this issue: #2237 is about adding angled gunports. This issue is about using the engine space as a rear-facing gunport. Doing so would require additional C++ code over and above that required to implement #2237.

**Comment by Amazinite** [comment]
>This issue is about using the engine space as a rear-facing gunport. Doing so would require additional C++ code over and above that required to implement #2237.

#2237 encompasses this issue, though. If #2237 was implemented and the code looked something like
`gun <x> <y> <firing angle>` (as Todd mentions [here](https://github.com/endless-sky/endless-sky/issues/2237#issuecomment-338480464)), then all one would need to do to create a gun that takes up engine space and fires out an engine is to do `gun <engine x> <engine y> 180` and make the gun take up engine space, something that can already be done with the current code and therefore does not require additional C++ code changes. Just change `"weapon capacity"` to `"engine capacity"` on a weapon to make it take up engine space.

**Comment by jostephd** [comment]
Yes, but every ship's `<engine x>` and `<engine y>` are different. An outfit that hardcodes the `<engine x> <engine y>` parameters of any particular ship wouldn't make sense to install in any other ship. The whole point of a code change is to be able to declare an outfit that declares itself as "I fire from the same x,y as the thrusters, wherever that is".

**Comment by Amazinite** [comment]
>Yes, but every ship's <engine x> and <engine y> are different. An outfit that hardcodes the <engine x> <engine y> parameters of any particular ship wouldn't make sense to install in any other ship. 

Gun and engine positions are defined by the ship, not by the outfit. 

```
ship "Aerie"
	sprite "ship/aerie"
	thumbnail "thumbnail/aerie"
	...
	engine -10 91
	engine 10 91
	gun -13 -79 "Heavy Laser"
	gun 13 -79 "Heavy Laser"
	turret 0 0 "Heavy Anti-Missile Turret"
	turret -17 14 "Heavy Laser Turret"
	turret 17 14 "Heavy Laser Turret"
	fighter -42 -2
	fighter 42 2
	...
```

**Comment by jostephd** [comment]
> Gun and engine positions are defined by the ship, not by the outfit.

True. In your example, the outfit declaration might look like:

```
outfit "Reverse gun port"
	"engine gunport" 1
	"engine space" -20
	"engine weapon capacity" 20
	description "Convert some engine space for use as a gunport."
```

When that outfit is installed in an Aerie, the game engine would add a `gun -10 91 180` gunport.
